[Site Isolation] Make tooltips work with site isolation.

megangardner opened the pull request:
#### a080b72f41e32c392811ef191c8c2a96e7560e38
<pre>
[Site Isolation] Make tooltips work with site isolation.
<a href="https://bugs.webkit.org/show_bug.cgi?id=296930">https://bugs.webkit.org/show_bug.cgi?id=296930</a>
<a href="https://rdar.apple.com/155247419">rdar://155247419</a>

Reviewed by Abrar Rahman Protyasha.

This is the fix and a test for tooltips to work correctly
with site isolation.

The fix is in EventHandler.cpp. In order to make sure that
mouse events get to the right process, currently first send
the event to the main frame web-process. If we are triggering in
a sub-frame, that information is passed onto a subframe in
the remoteUserInputEventData. If we have that data then we
know that the information is destined to head to another
process and we should not finish processing this
event in the main frame web process and returing to the
UI process. This check is all we need to circumvent
that.

Many of the rest of the changes are fixing capitalization of
toolTips to tooltips.

The rest is the plumbing needed to write a proper test.
Our current tooltip tests work by merely asking the web
process what tooltip we would expect for an element. That
is basically useless in a site isolated world. I have
added C API to allow for the passing back of the current
tooltip string when the tooltip changes. This is C API
only since the only use is for testing.

* LayoutTests/http/tests/site-isolation/mouse-events/iframes-with-tooltips-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mouse-events/iframes-with-tooltips.html: Added.
* LayoutTests/http/tests/site-isolation/mouse-events/resources/tooltip-subframe-a.html: Added.
* LayoutTests/http/tests/site-isolation/mouse-events/resources/tooltip-subframe-b.html: Added.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::mouseMoved):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::tooltipDidChange):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/C/WKPageUIClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::mouseDidMoveOverElement):
(WebKit::WebPageProxy::setToolTip):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::handleContextMenuLookUpImage):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::tooltipDidChange):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setTooltipDidChangeCallback):
(WTR::TestController::tooltipDidChange):
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::runTest):
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:

Canonical link: <a href="https://commits.webkit.org/298294@main">https://commits.webkit.org/298294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28de0370ac21711f231dbe356b98c4d7e86fc602

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34735 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25216 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121139 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65669 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1e90094d-b1bd-4423-b226-deb143d7d15a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116908 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43296 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87403 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f8f32f19-b5dc-4387-80f4-c5f6cf11b5a5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103265 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67799 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27367 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21385 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64791 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97576 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21501 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124329 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31393 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96201 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42364 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99455 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95986 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24430 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41185 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19027 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38004 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41867 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47402 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41411 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44728 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43148 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on LayoutTests/http/tests/site-isolation/mouse-events/iframes-with-tooltips.html:
```diff
@@ -0,0 +1,81 @@
+        }
+        {
+            let promiseClearing = testRunner.tooltipDidChange();
+            eventSender.asyncMouseMoveTo(100, 100);
```
not awaiting these async mouse moves is strange, is that intentional?

megangardner replied:
nope, mistake, thanks!

hortont424 commented on LayoutTests/http/tests/site-isolation/mouse-events/iframes-with-tooltips.html:
```diff
@@ -0,0 +1,81 @@
+            shouldBe("result", "''");
+
+            let promise = testRunner.tooltipDidChange();
+            await eventSender.asyncMouseMoveTo(300, 300);
```
this second mouse move seems wrong? the clear + set pair come inside a single mousemove, don't they?

megangardner replied:
Yeah, it was late and I didn't check it in the morning...

pxlcoder commented on LayoutTests/http/tests/site-isolation/mouse-events/iframes-with-tooltips.html:
```diff
@@ -0,0 +1,81 @@
+    </div>
+</body>
+
+
```
Stray newlines.

aprotyas commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2194,7 +2194,8 @@ HandleUserInputEventResult EventHandler::mouseMoved(const PlatformMouseEvent& ev
 
     hitTestResult.setToNonUserAgentShadowAncestor();
-    page->chrome().mouseDidMoveOverElement(hitTestResult, event.modifiers());
+    if (!result.remoteUserInputEventData())
+        page->chrome().mouseDidMoveOverElement(hitTestResult, event.modifiers());
```
Do we not need to return the `remoteMouseEventData` if `result.remoteMouseEventData.has_value()`?

megangardner replied:
We are returning the result, which has the remoteUserInputEventData on it, and that is already dealing with re-routing the mouse events and was a different set of work that is already done.

megangardner committed:
[Site Isolation] Make tooltips work with site isolation.
https://bugs.webkit.org/show_bug.cgi?id=296930
rdar://155247419

Reviewed by Abrar Rahman Protyasha.

This is the fix and a test for tooltips to work correctly
with site isolation.

The fix is in EventHandler.cpp. In order to make sure that
mouse events get to the right process, currently first send
the event to the main frame web-process. If we are triggering in
a sub-frame, that information is passed onto a subframe in
the remoteUserInputEventData. If we have that data then we
know that the information is destined to head to another
process and we should not finish processing this
event in the main frame web process and returing to the
UI process. This check is all we need to circumvent
that.

Many of the rest of the changes are fixing capitalization of
toolTips to tooltips.

The rest is the plumbing needed to write a proper test.
Our current tooltip tests work by merely asking the web
process what tooltip we would expect for an element. That
is basically useless in a site isolated world. I have
added C API to allow for the passing back of the current
tooltip string when the tooltip changes. This is C API
only since the only use is for testing.

* LayoutTests/http/tests/site-isolation/mouse-events/iframes-with-tooltips-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mouse-events/iframes-with-tooltips.html: Added.
* LayoutTests/http/tests/site-isolation/mouse-events/resources/tooltip-subframe-a.html: Added.
* LayoutTests/http/tests/site-isolation/mouse-events/resources/tooltip-subframe-b.html: Added.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::mouseMoved):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::tooltipDidChange):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/C/WKPageUIClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::mouseDidMoveOverElement):
(WebKit::WebPageProxy::setToolTip):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::handleContextMenuLookUpImage):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::tooltipDidChange):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setTooltipDidChangeCallback):
(WTR::TestController::tooltipDidChange):
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::runTest):
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:

Canonical link: https://commits.webkit.org/298294@main
